File: app/src/main/java/com/samourai/sentinel/service/WebSocketService.java
package com.samourai.sentinel.service;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.os.IBinder;

import com.samourai.sentinel.SamouraiSentinel;
import com.samourai.sentinel.util.ReceiveLookAtUtil;

import java.util.List;
import java.util.Timer;
import java.util.TimerTask;
import android.util.Log;

public class WebSocketService extends Service {

    private Context context = null;

    private Timer timer = new Timer();
    private static final long checkIfNotConnectedDelay = 15000L;
    private WebSocketHandler webSocketHandler = null;
    private final Handler handler = new Handler();
    private String[] addrs = null;

    public static List<String> addrSubs = null;

    @Override
    public void onCreate() {

        super.onCreate();

        //
        context = this.getApplicationContext();

        List<String> addrSubs = SamouraiSentinel.getInstance(WebSocketService.this).getAllAddrsSorted();
        addrs = addrSubs.toArray(new String[addrSubs.size()]);

        if(addrs.length == 0)    {
            return;
        }

        webSocketHandler = new WebSocketHandler(WebSocketService.this, addrs);
        connectToWebsocketIfNotConnected();

        timer.scheduleAtFixedRate(new TimerTask() {
            @Override
            public void run() {
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        connectToWebsocketIfNotConnected();
                    }
                });
            }
        }, 5000, checkIfNotConnectedDelay);

    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    public void connectToWebsocketIfNotConnected()
    {
        try {
            if(!webSocketHandler.isConnected()) {
                webSocketHandler.start();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void stop() {
        try {
            if(webSocketHandler != null)    {
                webSocketHandler.stop();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onDestroy()
    {
        stop();
        super.onDestroy();
    }

}

File: app/src/main/java/com/samourai/sentinel/service/WebSocketService.kt
package com.samourai.sentinel.service

import android.app.Service
import android.app.job.JobInfo
import android.app.job.JobParameters
import android.app.job.JobScheduler
import android.app.job.JobService
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.os.IBinder
import com.samourai.sentinel.api.ApiService
import org.koin.java.KoinJavaComponent
import org.koin.java.KoinJavaComponent.inject
import timber.log.Timber

class WebSocketService : JobService() {
    private val webSocketHandler: WebSocketHandler by inject(WebSocketHandler::class.java);

    override fun onStartJob(p0: JobParameters?): Boolean {
        Timber.i("onStartJob ")
        webSocketHandler.connect()
        return false
    }

    override fun onStopJob(p0: JobParameters?): Boolean {
        Timber.i("onStopJob")
        webSocketHandler.dispose()
        return true
    }

    override fun onDestroy() {
        super.onDestroy()
        Timber.i("onDestroy")
    }

    override fun onRebind(intent: Intent?) {
        super.onRebind(intent)
        Timber.i("onRebind ")
    }

    override fun onLowMemory() {
        super.onLowMemory()
        Timber.i("onLowMemory")
    }

    companion object {
        fun start(mContext: Context) {
            val mJobScheduler = mContext.getSystemService(Context.JOB_SCHEDULER_SERVICE) as JobScheduler
            val jobInfo = JobInfo.Builder(12, ComponentName(mContext, WebSocketService::class.java))
                    .setPersisted(true)
                    .setRequiredNetworkType(JobInfo.NETWORK_TYPE_ANY)
                    .build()

            mJobScheduler.schedule(jobInfo)
        }
    }
}
